CBinding/CMake/CMakeFileFormat.cs

```diff
@@ -0,0 +1,289 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Collections.Generic;
+
+using MonoDevelop.Core;
+using MonoDevelop.Projects;
+using MonoDevelop.Ide.Tasks;
+using MonoDevelop.Core.Logging;
+
+namespace CBinding {
+	public class CMakeFileFormat
+	{
+		//File path of the current CMakeLists.txt
+		public FilePath File;
+		public CMakeProject Project;
+		public string ProjectName = ""; //Comes from `project (project_name)` cmake command.
+		public Dictionary<string, CMakeCommand> SetCommands = new Dictionary<string, CMakeCommand> ();
+		public Dictionary<string, CMakeTarget> Targets = new Dictionary<string, CMakeTarget> ();
+		public Dictionary<FilePath, CMakeFileFormat> Children = new Dictionary<FilePath, CMakeFileFormat> ();
+		
+		public CMakeFileFormat Parent;
+		Dictionary<string, CMakeCommand> allCommands = new Dictionary<string, CMakeCommand> ();
+		Stack<string> blocks = new Stack<string> ();
+		MonoDevelop.Projects.Text.TextFile contentFile;
+		
+		//[A-Za-z_][A-Za-z0-9_]*	: ID
+		//\[=[0-9]*\[.*?\]=[0-9]*\]	: bracket argument.
+		string commandPattern = @"[A-Za-z_][A-Za-z0-9_]*\s*\((\"".*?\""|\[=[0-9]*\[.*?\]=[0-9]*\]|.)*?\)";
+		string blockCommentPattern = @"#\[\[.*?\]\]";
+		string lineCommentPattern = @"#.*?\n";
+
+		bool checkSyntax ()
+		{
+			var checkText = Regex.Replace (contentFile.Text, lineCommentPattern, "\n", RegexOptions.Singleline);
+			checkText = Regex.Replace (checkText, blockCommentPattern, "", RegexOptions.Singleline);
+			checkText = Regex.Replace (checkText, @"\s+", " ", RegexOptions.Singleline);
+			checkText = Regex.Replace (checkText, commandPattern, "", RegexOptions.Singleline);
+			if (checkText.Trim ().Length > 0) {
+				MonoDevelop.Ide.MessageService.ShowError ("Couldn't the load project.", String.Format ("Syntax errors found in {0}", File.ToString ()));
```

Comment:
This is UI code, which doesn't really belong in the file format code. It would mean it couldn't be used from mdtool, for example.

I would define an InvalidCMakeFileException and throw that, and have something higher up the call stack catch the exception and report to an IProgressMonitor - see e.g. the solution file loader https://github.com/mono/monodevelop/blob/9960f77b54271957cf8935fcdbae36d11b007336/main/src/core/MonoDevelop.Core/MonoDevelop.Projects.Formats.MSBuild/SlnFileFormat.cs#L631

It would also be good to display location info, if possible.